planner/core/find_best_task.go

```diff
@@ -418,6 +442,204 @@ func (ds *DataSource) findBestTask(prop *property.PhysicalProperty) (t task, err
+		partialCost += count * netWorkFactor
+		return indexSel, partialCost, count
+	}
+	return is, partialCost, rowCount
+}
+
+func (ds *DataSource) convertToPartialTableScan(prop *property.PhysicalProperty, path *accessPath) (PhysicalPlan, float64, float64) {
+	var partialCost float64
+	ts := PhysicalTableScan{
+		Table:           ds.tableInfo,
+		Columns:         ds.Columns,
+		TableAsName:     ds.TableAsName,
+		DBName:          ds.DBName,
+		isPartition:     ds.isPartition,
+		physicalTableID: ds.physicalTableID,
+	}.Init(ds.ctx)
+	ts.SetSchema(ds.schema)
+	if ts.Table.PKIsHandle {
+		if pkColInfo := ts.Table.GetPkColInfo(); pkColInfo != nil {
+			if ds.statisticTable.Columns[pkColInfo.ID] != nil {
+				ts.Hist = &ds.statisticTable.Columns[pkColInfo.ID].Histogram
+			}
+		}
+	}
+	ts.Ranges = path.ranges
+	ts.AccessCondition, ts.filterCondition = path.accessConds, path.tableFilters
+	rowCount := path.countAfterAccess
+	// Adjust number of rows we actually need to scan if prop.ExpectedCnt is smaller than the count we calculated.
+	if prop.ExpectedCnt < ds.stats.RowCount {
+		count, ok, corr := ds.crossEstimateRowCount(path, prop.ExpectedCnt, false)
+		if ok {
+			rowCount = count
+		} else if corr < 1 {
+			correlationFactor := math.Pow(1-corr, float64(ds.ctx.GetSessionVars().CorrelationExpFactor))
+			selectivity := ds.stats.RowCount / rowCount
+			rowCount = math.Min(prop.ExpectedCnt/selectivity/correlationFactor, rowCount)
+		}
+	}
+	ts.stats = property.NewSimpleStats(rowCount)
+	ts.stats.StatsVersion = ds.statisticTable.Version
+	if ds.statisticTable.Pseudo {
+		ts.stats.StatsVersion = statistics.PseudoVersion
+	}
+	// scan cost
+	partialCost += rowCount * scanFactor
+	// network cost
+	partialCost += rowCount * netWorkFactor
+	return ts, partialCost, rowCount
+}
+
+func (ds *DataSource) convertToTableScanForIndexMerge(prop *property.PhysicalProperty, path *accessPath, totalRowCount float64) (PhysicalPlan, float64) {
+	var partialCost float64
+	ts := PhysicalTableScan{
+		Table:           ds.tableInfo,
+		Columns:         ds.Columns,
+		TableAsName:     ds.TableAsName,
+		DBName:          ds.DBName,
+		isPartition:     ds.isPartition,
+		physicalTableID: ds.physicalTableID,
+	}.Init(ds.ctx)
+	ts.SetSchema(ds.schema)
+	if ts.Table.PKIsHandle {
+		if pkColInfo := ts.Table.GetPkColInfo(); pkColInfo != nil {
+			if ds.statisticTable.Columns[pkColInfo.ID] != nil {
+				ts.Hist = &ds.statisticTable.Columns[pkColInfo.ID].Histogram
+			}
+		}
+	}
+	partPathsConds := make([]expression.Expression, 0, 1)
+	for _, cond1 := range ds.pushedDownConds {
+		find := true
+		for _, cond2 := range path.tableFilters {
+			if cond1 == cond2 {
+				find = false
+				break
+			}
+		}
+		if find {
+			partPathsConds = append(partPathsConds, cond1)
+		}
+	}
+	// scan cost
+	partialCost += totalRowCount * scanFactor
+	if len(partPathsConds) > 0 {
+		partialCost += totalRowCount * cpuFactor
+		selectivity, _, err := ds.stats.HistColl.Selectivity(ds.ctx, partPathsConds)
```

Comment:
This looks incorrect, you don't need to call `Selectivity` again here, the input/output row count of table scan here is `totalRowCount`, the output row count of selection above this table scan should be same as `ds.stats.RowCount`, which has been calculated in `DeriveStats` using `Selectivity` already, so we should not modify stats of `ds` now IMHO.